### PR TITLE
[TPCC] Waiting for MYSQL container to be healthy before starting benchmark

### PR DESCRIPTION
**Pull request by CharlesBuell**
# sysbench bash script now checks for healthy container

1. Initial setup -- first pane is what is going to be ran, second is `podman stats`, third is a command ready to check that the `mysql1` container is running before killing it

image

2. During startup -- Benchmark script has started, and `podman stats` is showing that both sysbench and mysql containers are running. The last pane confirms that `mysql` is running.

image

3. `mysql1` container killed mid-startup -- The bottom pane has the kill command, and the logs show that we've stopped the sysbench script mid startup to prevent further painpoints.

image

# Why check?

This screenshot shows what happens without this implementation. Although the `mysql1` container no longer exists, sysbench is still trying to check for the logs.
image

### CODE REVIEW

**Comment by CharlesBuell** [comment]
New modification: All MySQL containers will be checked for their final state before continuing or exiting. If one container is unhealthy, it won't immediately exit without checking if other containers are healthy. Instead, it will be removed from an array of checked arrays, and the user will be told after the pause function which containers failed before returning 1.

1. Initial state -- In the first pane there is a new argument `-i 2`, which specifies how many instances we are going to start up. The default is 1, but in this case we're going to have 2 sysbench containers and 2 MySQL containers. First pane is the command we're going to run, second is `podman stats` and the third shows a command that we're going to run to see that the container `mysql2` is actually running before killing it in step 3.

image

2. Starting the benchmark -- We've started all containers, and they can all be seen in the second pane with `podman stats`. In the third pane we can also see that `mysql2` is confirmed to be running.

image

3. Observing functionality of the pause function -- `mysql2` is killed in the third pane, and can no longer be seen in `podman stats` within the second pane. In the first pane, there's instructions on what to do for a fresh run, along with which containers ended up in an unhealthy state. In this case we only see one, which is `mysql2`.

image

File: benchmarks/tpcc/sysbench-tpcc.sh [hunk 1]
```diff
@@ -529,22 +529,71 @@ function start_sysbench_containers()
...
+    if [ ${#unhealthy_containers[@]} -ne 0 ]; then
+        error_msg "The following containers are not healthy:"
+        for container in "${unhealthy_containers[@]}"; do
+            error_msg "- $container"
+        done
+        error_msg "In order to do another fresh run, do the following steps (only 1 instance as an example, this benchmark run had $PM_INSTANCES):"
+        error_msg "- Stop and remove all sysbench containers, ex. 'podman stop sysbench1 && podman rm sysbench1'"
+        error_msg "- Remove all mysql containers, ex. 'podman rm mysql1'"
+        error_msg "- Delete the data volume mounted for all mysql containers, ex. 'sudo rm -rf /data/mysql1'"
+        return 1
```
Comment (bhshankar) [inline benchmarks/tpcc/sysbench-tpcc.sh:589]: At this point should we get_container_logs and stop_containers?  the mysql* will be stopped and the sysbench* will be killed.

We will have capture any logs that will be needed to debug the problem.

**Comment by CharlesBuell** [comment]
New functionality -- Stop and remove healthy containers and remove all data at `MYSQL_DATA_DIR` on detection of an unhealthy container

image

File: benchmarks/tpcc/sysbench-tpcc.sh (hunk 1 above)
Comment (CharlesBuell) [reply benchmarks/tpcc/sysbench-tpcc.sh:589]: Done in [c7787d3](https://github.com/cxlbench/cxlbench/pull/28/commits/c7787d30e8d0527f25c55af5d1c8edbd90ba7e14)

File: benchmarks/tpcc/sysbench-tpcc.sh [hunk 2]
```diff
@@ -1075,6 +1121,18 @@ function remove_containers()
...
 
+# Remove the MySQL volume data located at /data/mysql${i}
+# args: none
+# return: nothing
+function delete_container_data()
+{
+    for i in $(seq 1 ${PM_INSTANCES});
+    do
+        info_msg "Removing mysql${i} data at $MYSQL_DATA_DIR/mysql${i}"
+        sudo rm -rf $MYSQL_DATA_DIR/mysql${i}
```
Comment (bhshankar) [inline benchmarks/tpcc/sysbench-tpcc.sh]: ${MYSQL_DATA_DIR} please

File: benchmarks/tpcc/sysbench-tpcc.sh [hunk 3]
```diff
@@ -1622,6 +1680,11 @@ for function in "${functions[@]}"; do
 
     # Check if an error occurred
     if [ $return_value -ne 0 ]; then
+        get_container_logs
+        stop_containers
+        remove_containers
+        delete_container_data
+
         error_msg "An error occurred in '$function'. Exiting."
```
Comment (bhshankar) [inline benchmarks/tpcc/sysbench-tpcc.sh:1688]: Please add
error_msg "... See container logs at '${OUTPUT_PATH}/${OUTPUT_PREFIX}mysql.*.log' for details"

Comment (CharlesBuell) [reply benchmarks/tpcc/sysbench-tpcc.sh:1688]: 1497db9afb22

File: benchmarks/tpcc/sysbench-tpcc.sh (hunk 2 above)
Comment (CharlesBuell) [reply benchmarks/tpcc/sysbench-tpcc.sh]: 1497db9afb22

**Review by bhshankar — APPROVED** [review]

**Pull request merged**